lib/core-server/src/utils/watch-story-specifier.ts

```diff
@@ -0,0 +1,47 @@
+import Watchpack from 'watchpack';
+import { toRequireContext, NormalizedStoriesSpecifier } from '@storybook/core-common';
+import { Path } from '@storybook/store';
+
+// TODO -- deal with non "specified" specifiers
+export function watchStorySpecifiers(
+  specifiers: NormalizedStoriesSpecifier[],
+  onInvalidate: (specifier: NormalizedStoriesSpecifier, path: Path, removed: boolean) => void
+) {
+  // See https://www.npmjs.com/package/watchpack for full options.
+  // If you want less traffic, consider using aggregation with some interval
+  const wp = new Watchpack({
+    // poll: true, // Slow!!! Enable only in special cases
+    followSymlinks: false,
+    ignored: ['**/.git', 'node_modules'],
+  });
+  console.log(specifiers.map((ns) => ns.specifier.directory));
+  wp.watch({
+    directories: specifiers.map((ns) => ns.specifier.directory),
+  });
+
+  function onChangeOrRemove(path: Path, removed: boolean) {
+    console.log('onChangeOrRemove', path, removed);
+    const specifier = specifiers.find((ns) => {
+      const { path: base, regex } = toRequireContext(ns.glob);
+      console.log(
+        base,
+        regex,
+        !!path.startsWith(base.replace(/^\.\//, '')),
+        !!path.match(new RegExp(regex))
+      );
+      return path.startsWith(base.replace(/^\.\//, '')) && path.match(new RegExp(regex));
```

Comment:
This `./` seems a little dicey? Would be nice to figure this out globally somehow so we don't have to keep special casing things everywhere

Reply:
I agree. Consolidating the specifier code (with the require context and import fn code also) is at the top of my list here.